File: src/sorting/cocktailShakerSort.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include "./_common_/common.h"

void cocktail_sort(int* arr, const int size)
{
    int half = size / 2;
    bool is_swapped = false;
    
    for (int i = 0; i < half; i++)
    {
        for (int j = i; j < size - i - 1; j++)
        {
            if (arr[j] > arr[j + 1])
            {
                swap(&arr[j], &arr[j + 1]);
                is_swapped = true;
            }
        }
        for (int j = size - i - 2; j > i; j--)
        {
            if (arr[j - 1] > arr[j])
            {
                swap(&arr[j - 1], &arr[j]);
                is_swapped = true;
            }
        }
        if (!is_swapped) break;
    }
}

int main(void)
{
    int n;

    scanf("%d", &n);
    int* arr = (int*)malloc(sizeof(int) * n);
    for (int i = 0; i < n; i++)
        scanf("%d", &arr[i]);

    print_array(arr, n);
    cocktail_sort(arr, n);
    print_array(arr, n);

    free(arr);
    return 0;
}

/*
Input:
5
5 4 3 2 1

Output:
5 4 3 2 1
1 2 3 4 5
*/

File: src/sorting/cocktailShakerSort.py
def cocktail_sort(arr, size):
    r = size // 2
    is_swapped = False

    for i in range(r):
        for j in range(i, size - i - 1):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                is_swapped = True
        for j in range(size - i - 2, i, -1):
            if arr[j - 1] > arr[j]:
                arr[j - 1], arr[j] = arr[j], arr[j - 1]
                is_swapped = True
        if is_swapped == False:
            break


def print_array(arr, size):
    for i in range(size):
        print(arr[i], end=' ')
    print()


arr = list(map(int, input().split()))
size = len(arr)

print_array(arr, size)
cocktail_sort(arr, size)
print_array(arr, size)
